### PR TITLE
Fix a COMDouble::Round() issue

### PR DESCRIPTION
**Pull request by hanblee**
fixes https://github.com/dotnet/coreclr/issues/12137

### CODE REVIEW

**Comment by hanblee** [comment]
@tannergooding PTAL

**Comment by hanblee** [comment]
@tannergooding BTW, if you are not the right person to review this, it would be great if you would please pull in the right person(s). Thanks!

**Comment by tannergooding** [comment]
Taking a look now.

**Comment by tannergooding** [comment]
It is important to note that the MSDN [page](https://msdn.microsoft.com/en-us/library/wyk4d9cy(v=vs.110).aspx) for `Math.Round(double)` states:
> Rounds a double-precision floating-point value to the nearest integral value.

However, the actual behavior, which is on the general `Math.Round` [page](https://msdn.microsoft.com/en-us/library/system.math.round(v=vs.110).aspx) is:
> Rounds a double-precision floating-point value to the nearest integer, and rounds midpoint values to the nearest even number 

The secondary portion (rounding midpoint values to the nearest even number) is important for just this scenario and the exact behavior taken here.

Additionally, as an FYI, the original (unmodified) code from Desktop was https://github.com/hanblee/coreclr/commit/08786f20e89eb5f518d8d25f3e7f886f69d994ea#diff-884cab80438864e5dc66c7e4c1a3c723L467, which is identical behavior to implementation currently checked in.

When completely "unrolled" the current implementation is doing
```C++
(floor(x + 0.5) == (x + 0.5)) && (fmod((x + 0.5), 2.0) != 0)
```
and the newly proposed implementation is doing:
```C++
(x == (floor(x) + 0.5)) && (fmod(floor(x + 0.5), 2.0) != 0)
```

The previous algorithm would end up comparing "precisely 1" vs "precisely 1".
The new algorithm ends up comparing 0.5 vs 0.50000000000000011, which is what allows this fix.

This is definitely an improvement as it should only do the midpoint rounding calculation when x is exactly midpoint (as best as can be represented by IEEE double-precision floating-point values, anyways).

**Comment by tannergooding** [comment]
It would be great if you could add a couple tests covering this scenario (CoreFX is the primary repo for managed tests, but there are some tests in CoreCLR as well: https://github.com/dotnet/coreclr/tree/32f0f9721afb584b4a14d69135bea7ddc129f755/tests/src/CoreMangLib/cti/system/math)

**Comment by tannergooding** [comment]
FYI. @janvorli, @AndyAyersMS, @jkotas who have reviewed these types of changes previously.

**Comment by hanblee** [comment]
> It would be great if you could add a couple tests covering this scenario 

Good point. Will do.

File: tests/src/CoreMangLib/cti/system/math/mathround3.cs [hunk 1]
```diff
@@ -91,6 +93,48 @@ public bool PosTest2()
...
+
+        // Test based on https://github.com/dotnet/coreclr/issues/12137
+        TestLibrary.TestFramework.BeginScenario("PosTest4: Verify Round(System.double) when decimal part of arg is very close to 0.5 .");
+
+        double doubleVal = 0.50000000000000011102230246251565404236316680908203;
+        double expectedVal = 1.0;
+
+        if (Math.Round(doubleVal) != expectedVal)
+        {
+            Console.WriteLine("actual value = " + Math.Round(doubleVal));
```
Comment (tannergooding) [inline tests/src/CoreMangLib/cti/system/math/mathround3.cs]: Could you use the `G17` format specifier when printing here and below?
```
> double doubleVal = 0.50000000000000011102230246251565404236316680908203;
> Console.WriteLine(doubleVal)
0.5
> Console.WriteLine("{0:G17}", doubleVal)
0.50000000000000011
```

Comment (hanblee) [inline tests/src/CoreMangLib/cti/system/math/mathround3.cs]: Done

**Comment by tannergooding** [comment]
Thanks! This will need someone from the CoreCLR team to give final approval before it can be merged.

**Comment by hanblee** [comment]
Got it!

**Comment by hanblee** [comment]
@janvorli PTAL

**Pull request merged**